File: 프로그래머스/lv2/87946. 피로도/피로도.kt
class Solution {
    var ans = 0
    
    fun solution(k: Int, dungeons: Array<IntArray>): Int {
        backtracking(k, dungeons, BooleanArray(dungeons.size) { false }, 0)
        
        return ans
    }
    
    fun backtracking(k: Int, dungeons: Array<IntArray>, visited: BooleanArray, cnt: Int) {
        if (cnt > ans) {
            ans = cnt
        }
        
        dungeons.forEachIndexed { idx, it ->
            if (visited[idx] == false && k >= it[0]) {
                visited[idx] = true
                backtracking(k - it[1], dungeons, visited, cnt + 1)
                visited[idx] = false
            }
        }
    }
}

File: 프로그래머스/lv2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = 0
    
    for i in permutations(dungeons, len(dungeons)):
        hp = k
        cnt = 0
        
        for j in i:
            if hp >= j[0]:
                hp -= j[1]
                cnt += 1
                
        if answer < cnt:
            answer = cnt
    
    return answer